lib/krb5/crypto.c

```diff
@@ -472,6 +472,7 @@ create_checksum_iov(krb5_context context,
 		    unsigned usage,
 		    struct krb5_crypto_iov *iov,
 		    int niov,
+		    krb5_flags flags,
```

Comment:
Looking at these static create/verify functions, they a,ll seem to be handed the crypto object, do we really need to change their signature, can't they just use crypto_flags(crypto)?

Reply:
We always want to require keyed checksums when `krb5_encrypt()` (and friends) are used, at least for non-DES encryption types. `KRB5_CRYPTO_FLAG_ALLOW_UNKEYED_CHECKSUM` should only apply to `krb5_{create, verify}_checksum()`, not the encryption/decryption functions.

Reply:
i.e. `crypto_flags(crypto)` is honoured only when we're dealing with the checksum API functions (not the encryption ones)

Reply:
Oh I see, that explain the changes in the encryption functions, starting to get it.

Reply:
So, `encrypt_internal()` is used (as far as I can tell!) by DES only, so we explicitly set `KRB5_CRYPTO_FLAG_ALLOW_UNKEYED_CHECKSUM` there. Otherwise, `encrypt_internal_derived()` is used for 3DES and AES-SHA1, `encrypt_internal_special()` for RC4, and `encrypt_internal_enc_then_cksum()` for AES-SHA2. These latter functions do not set `KRB5_CRYPTO_FLAG_ALLOW_UNKEYED_CHECKSUM`, although that was never a vulnerability because they explicitly use `et->keyed_checksum` (i.e. the mandatory keyed checksum type is used for the encryption type). Not setting `KRB5_CRYPTO_FLAG_ALLOW_UNKEYED_CHECKSUM` is just good hygiene / defense in depth.

Reply:
Got it, fwiw it all looks good to me, thanks!

Reply:
Thanks for the review! Merged :)